clang/lib/CIR/CodeGen/CIRGenStmt.cpp

```diff
@@ -383,6 +383,16 @@ mlir::LogicalResult CIRGenFunction::buildIfStmt(const IfStmt &S) {
       if (buildStmt(S.getInit(), /*useCurrentScope=*/true).failed())
         return mlir::failure();
 
+    // Handle "if constexpr" explicitly here to avoid generating some
+    // ill-formed code since in CIR the "if" are no longer simplified
+    // in this lambda like in Clang but postponed to other MLIR
+    // passes.
+    if (S.isConstexpr())
+      if (auto ndc = S.getNondiscardedCase(getContext())) {
+        // Replace the "if constexpr" by its non-discarded branch
+        return buildStmt(ndc.value(), /*useCurrentScope=*/true);
+      }
+
```

Comment:
The comment I left on the code is talking about early optimizations done in LLVM codegen that aren't necessarily constexpr related. I believe we can still follow the same skeleton. For instance, instead of adding this here, have you tried to implement this inside the check below:

```
    if (ConstantFoldsToSimpleInteger(S.getCond(), CondConstant,
                                     S.isConstexpr())) {
     // In CIR we don't flip the conditions right away
     if (S.isConstexpr()) {
        RunCleanupsScope Scope(*this);
        buildStmt(S);
      }
      ...
     }
      assert(!UnimplementedFeature::constantFoldsToSimpleInteger());
    }
```

If you did, what didn't work?

Reply:
Note that expensive checker is still useful because we can track of that missing optimizations because of the `assert(!UnimplementedFeature::constantFoldsToSimpleInteger());`. Once we have a passa that implements this, we could use your thiner version (just leave guidance in the comments).

Reply:
Yes, it works but the current code calling the `ifStmtBuilder` lambda always creates a `cir.scope` while Clang can avoid generating anything in that case.
Cf https://github.com/keryell/clangir/commit/a547eab57380f64c77b6acdc624de017888b8974#diff-0cb68c7505de027f386b8b1c8a4bfdf4b2f26978d71c8a52b3447116715ef7f5R77 generated from https://github.com/keryell/clangir/commit/a547eab57380f64c77b6acdc624de017888b8974#diff-0cb68c7505de027f386b8b1c8a4bfdf4b2f26978d71c8a52b3447116715ef7f5R31
Is this OK or do I need to rework the logic of `ifStmtBuilder` and its caller?

Reply:
This is fine, the `cir.scope` is created to represent the overall scope where the if codegen actually happens within (it's where init statements will place their alloca, etc). Poking at the IfStmt some layers above for this optimization isn't worth it IMO - the cleanup pass should be stripping the `cir.scope` here anwyays, but let me track that as a separate issue.

Thanks for improving the PR, LGTM!